src/hotspot/os/posix/signals_posix.cpp

```diff
@@ -76,22 +99,68 @@ extern "C" {
   typedef void (*sa_sigaction_t)(int, siginfo_t *, void *);
 }
 
-// For diagnostics to print a message once. see run_periodic_checks
-static sigset_t check_signal_done;
-static bool check_signals = true;
+// At various places we store handler information for each installed handler.
+//  SavedSignalHandlers is a helper class for those cases, keeping an array of sigaction
+//  structures.
+class SavedSignalHandlers {
+  // Note: NSIG can be largish, depending on platform, and this array is expected
+  // to be sparsely populated. To save space the contained structures are
+  // C-heap allocated. Since they only get added outside of signal handling
+  // this is no problem.
+  struct sigaction* _sa[NSIG];
+
+  bool check_signal_number(int sig) const {
+    assert(sig > 0 && sig < NSIG, "invalid signal number %d", sig);
+    return sig > 0 && sig < NSIG;
+  }
+
+public:
+
+  SavedSignalHandlers() {
+    ::memset(_sa, 0, sizeof(_sa));
+  }
+
+  ~SavedSignalHandlers() {
+    for (int i = 0; i < NSIG; i ++) {
+      FREE_C_HEAP_OBJ(_sa[i]);
+    }
+  }
+
+  void set(int sig, const struct sigaction* act) {
+    if (check_signal_number(sig)) {
+      assert(_sa[sig] == NULL, "Overwriting signal handler?");
+      _sa[sig] = NEW_C_HEAP_OBJ(struct sigaction, mtInternal);
+      *_sa[sig] = *act;
+    }
+  }
+
+  const struct sigaction* get(int sig) const {
+    if (check_signal_number(sig)) {
+      return _sa[sig];
+    }
+    return NULL;
+  }
+};
+
 
 debug_only(static bool signal_sets_initialized = false);
 static sigset_t unblocked_sigs, vm_sigs, preinstalled_sigs;
-struct sigaction sigact[NSIG];
 
-// For signal-chaining
+// Our own signal handlers should never ever get replaced by a third party one.
+//  To check that, and to aid with diagnostics, store a copy of the handler setup
+//  and compare it periodically against reality (see os::run_periodic_checks()).
+static bool check_signals = true;
+static SavedSignalHandlers expected_handlers;
```

Comment:
I personally would prefer `vm_handlers` or `java_handlers` or `our_handlers` here instead of `expected_handlers`.

Reply:
Okay, vm_handlers it is.